src/generic/Matrix.jl

```diff
@@ -184,13 +184,13 @@ ncols(a::MatrixElem) = size(a.entries, 2)
     length(a::Generic.MatrixElem)
 > Return the number of entries in the given matrix.
 """
-length(a::MatrixElem) = nrows(a) * ncols(a)
+length(a::AbstractAlgebra.MatElem) = nrows(a) * ncols(a)
```

Comment:
Actually, `AbstractAlgebra.MatElem <: Generic.MatrixElem`. With this change, if I'm not mistaken, you make `length` applicable to less types (you basically remove `length(::MatAlgElem)`)

Reply:
No, I don't think this is true.
```julia
julia> supertype(fmpz_mat)
MatElem{fmpz}

julia> supertype(Generic.Mat{fmpz})
MatElem{fmpz}

julia> supertype(MatElem{fmpz})
ModuleElem{fmpz}
```

Reply:
Sorry, I am confused.

Reply:
Indeed, but `MatrixElem` is `MatrixElem = Union{MatAlgElem{T}, MatElem{T}}`, so it doesn't appear via a `supertype` call. I needed to write down a graph with these types (with `<:` relationship) to understand better how they relate.

Reply:
I wrote this comment before seeing your last answer, I didn't reply "indeed" to you feeling confused!

Reply:
@rfourquet Could you make the graph and the script somewhere available? I'd be best to include it into documentation at some point.

Reply:
I did it first by hand (just the the fiew types under `MatrixElem`), but I have indeed started to work on a script, I will share when it's a bit less buggy (the Julia type system is quite complex for this task!)

Reply:
Alright, for this small example here is a preview: in red are types from `AbstractAlgebra.Generic`, in blue from `AbstractAlgebra`. Dashed borders means unexported, bold means abstract type. 
![Uploading matrix.png…]()

Reply:
looks good! is there a reason why there is `Mat{T}`? or the graph is incomplete?

Reply:
> is there a reason why there is Mat{T}?

What is wrong with it?